embassy-stm32/src/can/fdcan.rs

```diff
@@ -1,40 +1,640 @@
+    /// frame is dropped from the mailbox, it is returned.  If no lower-priority frames
+    /// can be replaced, this call asynchronously waits for a frame to be successfully
+    /// transmitted, then tries again.
+    pub async fn write(&mut self, frame: &TxFrame) -> Option<TxFrame> {
+        poll_fn(|cx| {
+            T::state().tx_waker.register(cx.waker());
+            if let Ok(dropped) =
+                self.can
+                    .borrow_mut()
+                    .transmit_preserve(frame.header, &frame.data.bytes, &mut |_, hdr, data32| {
+                        TxFrame::from_preserved(hdr, data32)
+                    })
+            {
+                return Poll::Ready(dropped.flatten());
+            }
+
+            // Couldn't replace any lower priority frames.  Need to wait for some mailboxes
+            // to clear.
+            Poll::Pending
+        })
+        .await
+    }
+}
+
+/// FDCAN Rx only Instance
+#[allow(dead_code)]
+pub struct FdcanRx<'c, 'd, T: Instance, M: fdcan::Receive> {
+    can: &'c RefCell<fdcan::FdCan<FdcanInstance<'d, T>, M>>,
+}
+
+impl<'c, 'd, T: Instance, M: fdcan::Receive> FdcanRx<'c, 'd, T, M> {
+    /// Returns the next received message frame
+    pub async fn read(&mut self) -> Result<RxFrame, BusError> {
+        poll_fn(|cx| {
+            T::state().err_waker.register(cx.waker());
+            T::state().rx_waker.register(cx.waker());
+
+            // TODO: handle fifo0 AND fifo1
+            let mut buffer: [u8; 64] = [0; 64];
+            if let Ok(rx) = self.can.borrow_mut().receive0(&mut buffer) {
+                // rx: fdcan::ReceiveOverrun<RxFrameInfo>
+                // TODO: report overrun?
+                //  for now we just drop it
+                let frame: RxFrame = RxFrame::new(rx.unwrap(), &buffer);
+                return Poll::Ready(Ok(frame));
+            } else if let Some(err) = self.curr_error() {
+                // TODO: this is probably wrong
+                return Poll::Ready(Err(err));
+            }
+
+            Poll::Pending
+        })
+        .await
+    }
+
+    fn curr_error(&self) -> Option<BusError> {
+        let err = { T::regs().psr().read() };
+        if err.bo() {
+            return Some(BusError::BusOff);
+        } else if err.ep() {
+            return Some(BusError::BusPassive);
+        } else if err.ew() {
+            return Some(BusError::BusWarning);
+        } else {
+            cfg_if! {
+                if #[cfg(stm32h7)] {
+                    let lec = err.lec();
+                } else {
+                    let lec = err.lec().to_bits();
+                }
+            }
+            if let Ok(err) = LastErrorCode::try_from(lec) {
+                return BusError::try_from(err);
+            }
+        }
+        None
+    }
+}
+impl<'d, T: Instance, M: FdcanOperatingMode> Deref for Fdcan<'d, T, M> {
+    type Target = RefCell<fdcan::FdCan<FdcanInstance<'d, T>, M>>;
```

Comment:
derefing to a RefCell is a bit risky, because it makes too easy to run into borrow errors together with async. For example, rx task borrows the refcell then awaits something else while holding it, then the tx task gets polled and tries to borrow the refcell again which fails.

If you use fdcan's `split_by_ref` you can make FdcanTx/FdcanRx own the Tx/Rx halves without having to use RefCells. Then Fdcan can deref to the full `fdcan::FdCan` (and maybe you want the tx/rx halves to deref to the fdcan Tx/Rx structs maybe?)

Reply:
I'll have a play with this. As you might have seen me mention, I'm pretty new to rust so these type of changes result in me fighting with the compiler quite a bit :) I'll use bxcan as a reference.

Reply:
I actually noticed yesterday when trying to add some tests after a split (feature/fdcan_experimental branch) that I could still access the non-split version after the split.

Reply:
Ok, that was easier than I thought it would be. There was less random trying of mut's and &' before and after :'s than I expected.